File: Backend/controllers/assignmentController.js
const Assignment = require('../models/Assignment')

// create a new assignment
exports.createAssignment = async (req, res) => {

    const { title, description, dueDate } = req.body

    try {
        // create new assignment
        const newAssignment = new Assignment({
            title: title,
            description: description, 
            datePosted: Date(),
            dueDate: dueDate
        })

        await newAssignment.save()

        return res.status(201).json({ message: 'Assignment created successfully' })
    } catch (error) {
        // catch server error
        console.error(error)
        return res.status(500).json({ message: 'Internal Server Error' })
    }
}

// get assignment by id
exports.getAssignmentById = async (req, res) => {
    const assignmentId = req.params.id

    try {
        const retrievedAssignment = await Assignment.findById(assignmentId)

        if (!retrievedAssignment) {
            // 404 if assignment with assignmentId does not exist
            return res.status(404),json({ error: 'Assignment not found' })
        } else {
            // return 200 and assignment if successful
            return res.status(200).json({ message: 'Assignment retrieved successfully', retrievedAssignment})
        }
    } catch (error) {
        // catch server error
        console.error(error)
        return res.status(500).json({ error: 'Internal Server Error' })
    }

}

// update assignment
exports.updateAssignment = async (req, res) => {
    const assignmentId = req.params.id
    const { title, description, dueDate} = req.body

    try {
        // find the assignment by id and update fields
        const updatedAssignment = await Assignment.findByIdAndUpdate(assignmentId, { title: title, description: description, dueDate: dueDate }, { new: true })

        if (!updatedAssignment) {
            // 404 if assignment with given id does not exist
            return res.status(404).json({ error: 'Assignment not found' })
        }

        // on success return 200 status, success message, and updated assignment
        res.status(200).json({ message: 'Assignment updated successfully', updatedAssignment })
    } catch (error) {
        // catch server error
        console.error(error)
        res.status(500).json({ error: 'Internal Server Error' })
    }
}

// delete assignment
exports.deleteAssignment = async (req, res) => {
    const assignmentId = req.params.id

    try {
        // find the assignment by id and delete
        const deletedAssignment = await Assignment.findByIdAndDelete(assignmentId)

        if (!deletedAssignment) {
            // 404 if assignment with given id does not exist
            return res.status(404).json({ error: 'Assignment not found' })
        }

        res.status(200).json({ message: 'Assignment deleted successfully' })
    } catch (error) {
        // catch server error
        console.error(error)
        res.status(500).json({ error: 'Internal Server Error' })
    }
}

File: Middle_School_Management_App/Middle_School_Management_App/AssignmentController.swift
//
//  AssignmentController.swift
//  Middle_School_Management_App
//
//  Created by Vivianne Yee on 2023-11-12.
//

import Foundation

class AssignmentController {

    init() {}
    
    enum NetworkError: Error {
        case emptyResponse
        case invalidResponse
        case serverError(String)
    }
    
    // create a new assignment
    func createAssignment(title: String, description: String, dueDate: Date, completion: @escaping (Result<Data, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/assignment/assignments")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")

                
        let body: [String: Any] = [
            "title": title,
            "description": description,
            "dueDate": dueDate
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body)
            request.httpBody = jsonData
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not create assignment")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 201 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            completion(.success(responseData))
        }.resume()
    }
    
    // get an assignment by its id
    func getAssignmentById(id: String, completion: @escaping (Result<Assignment, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/assignment/assignments/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not retrieve assignment")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let assignmentObject = try decoder.decode(Assignment.self, from: responseData)
                completion(.success(assignmentObject))
            } catch {
                completion(.failure(error))
            }
                        
        }.resume()
    }
    
    // update an assignment
    func updateAssignment(id: String, title: String, description: String, dueDate: Date, completion: @escaping (Result<Assignment, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/assignment/assignments/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "PUT"
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        
        let body: [String: Any] = [
            "title": title,
            "description": description,
            "dueDate": dueDate
        ]
        
        do {
            let jsonData = try JSONSerialization.data(withJSONObject: body)
            request.httpBody = jsonData
        } catch {
            completion(.failure(error))
            return
        }
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not update assignment")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            do {
                let decoder = JSONDecoder()
                let assignmentObject = try decoder.decode(Assignment.self, from: responseData)
                completion(.success(assignmentObject))
            } catch {
                completion(.failure(error))
            }
            
        }.resume()
    }
    
    // delete an assignment
    func deleteAssignment(id: String, completion: @escaping (Result<Data, Error>) -> Void) {
        let url = URL(string: "http://localhost:3000/assignment/assignments/\(id)")!
        var request = URLRequest(url: url)
        request.httpMethod = "DELETE"
        
        URLSession.shared.dataTask(with: request) { data, response, error in
            // handle error
            if let error = error {
                completion(.failure(error))
                print("Could not delete assignment")
                return
            }
            
            // check for successful response status
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                completion(.failure(NetworkError.invalidResponse))
                return
            }
            
            // check that response data is not empty
            guard let responseData = data else {
                completion(.failure(NetworkError.emptyResponse))
                return
            }
            
            completion(.success(responseData))
            
        }.resume()
    }
}
